implement dns service discovery, fix py_ipa_api_recreate script

schlitzered opened the pull request:
Hi, me again.

i implemented DNS service discovery.

the old  behavior will not change, but now you can simply omit setting a host, and the client will automatically try to do DNS service discovery., if the host running the tool, has no domainname, you can pass it as the following argument: dns_discovery="example.com"

i also fixed some errors in the tool used to recreate the client_meta.py file:
 - version has to be set using "version = self.json_spec['result']['messages'][0]['data']['server_version']", using self.json_spec['version'] is simply wrong. the later one is the version number of the IPA server, the first one is the actual version of the API
- some things got broken when adopting textwrap, i implemented a fallback to what has been done before, reason is that sometimes "spec['doc']" is None
- i also moved the code completely to the contrib, because a second run, after creating a broken file, would prevent the tool from running a second time (import error, because of the broken ClientMeta class)

i also did some consistency cleanups in the doc strings.

ClientMeta uses ":type  $PARAM: str", as it has been requested by you.

but Client, and ClientLegacy used ":type  $PARAM: string"
, so i changed it, to also use "str" instead of "string".


Kind Regards

schlitzered committed:
implement dns service discovery, fix py_ipa_api_recreate script

schlitzered committed:
bigfix, return session, when logging in using hostname

schlitzered committed:
fix test, sinse the test is not acutally login in to anything, we have to manually set self._current_host to what is expected.

schlitzered commented:
BTW ClientMeta is currently broken because  version is set to  '4.8.2', the IPA server will raise an error.

AmbientLighter commented on src/python_freeipa/client_meta.py:
```diff
@@ -25960,7 +25526,6 @@ def vault_retrieve_internal(
             o_raw=False,
     ):
         """
-        Retrieve data from a vault.
```
Hm, why this line is removed?

schlitzered commented:
hi, i just noticed that you cannot rely on the "class" attribute of an option. sometimes it is combined with an attribute that is called "multivalue", which basically tells that the options expects an array with strings in it.

i will fix this soon.

but taking into account that there is this issue with missing descriptions and so on, i would suggest to revoke the 1.0.0 release.

so we can fix and test the meta client carefully.

schlitzered commented:
and update here?

could you please let me know what version of IPA/IdM you are using?

i just checked both, RedHat Idm and Freeipa 4.6.5, and both do not seem to have the documentation string "Retrieve data from a vault." in there API spec.

AmbientLighter commented:
@schlitzered I'm using FreeIPA demo. Here's an example JSON response. As you can see it doesn't contain API version. Therefore I'd suggest to remove that line which fetches version from JSON,

https://gist.github.com/AmbientLighter/8ae2d0ed283eade9f1854305b4b77b54

AmbientLighter commented:
Also I have fixed API version string in meta client.

schlitzered commented:
interesting, the 4.6.5 version contains an attribute called "messages", which contains a message with the version number of the server.

schlitzered committed:
fix api doc, in case of "multivalue", which means the parameter accepts a list of strings
enhance py_ipa_api_recreate, to be able to fetch api definition and version from remote ipa server

schlitzered commented:
more changes:

i enhanced the py_ipa_api_recreate script, to allow to fetch the version and api spec from a remote server.

this eliminates the need to manually download the api spec.

it is still possible to use a local json file, but now you also have to specify the API version that the file uses. since the is no reliable way to fetch it from the schema spec itself.

i also fixed the api doc,in case of "multivalue": true, which means that the methods parameter expects a list of strings.

schlitzered committed:
add documentation on how to recreate ClientMeta class, and create custom package
